### PR TITLE
add KernelGradients

### PR DESCRIPTION
**Pull request by vchuravy**
# TODO:
- [x] Tests
- [x] https://github.com/JuliaGPU/AMDGPU.jl/pull/145

### CODE REVIEW

File: lib/KernelGradients/src/cuda.jl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import .CUDAKernels: CUDACtx, CUDACTX
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::CUDACtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(CUDACTX, f, args...))
+    Enzyme.autodiff_no_cassette(f′, args...)
+end
```
Comment (vchuravy) [inline lib/KernelGradients/src/cuda.jl:8]:
```suggestion
end

```

File: lib/KernelGradients/src/roc.jl [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import .ROCKernels: ROCCTX, ROCCtx
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::ROCCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(ROCCTX, f, args...))
+    Enzyme.autodiff_no_cassette(f′, args...)
+end
```
Comment (vchuravy) [inline lib/KernelGradients/src/roc.jl:8]:
```suggestion
end

```

File: lib/KernelGradients/src/roc.jl [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import .ROCKernels: ROCCTX, ROCCtx
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::ROCCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(ROCCTX, f, args...))
+    Enzyme.autodiff_no_cassette(f′, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/roc.jl:7]:
```suggestion
    Enzyme.autodiff(f′, args...)
```

File: lib/KernelGradients/src/cuda.jl [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import .CUDAKernels: CUDACtx, CUDACTX
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::CUDACtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(CUDACTX, f, args...))
+    Enzyme.autodiff_no_cassette(f′, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/cuda.jl:7]:
```suggestion
    Enzyme.autodiff(f′, args...)
```

File: lib/KernelGradients/src/KernelGradients.jl [hunk 5]
```diff
@@ -0,0 +1,25 @@
+module KernelGradients
+
+import KernelAbstractions: Kernel, CPUCTX, CPUCtx
+import Cassette
+import Enzyme
+using Requires
+
+@inline function Cassette.overdub(::CPUCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(CPUCTX, f, args...))
+    Enzyme.autodiff_no_cassette(f′, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/KernelGradients.jl:10]:
```suggestion
    Enzyme.autodiff(f′, args...)
```

File: lib/KernelGradients/Project.toml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+[deps]
+Cassette = "7057c7e9-c182-5462-911a-8362d720325c"
+Enzyme = "7da242da-08ed-463a-9acd-ee780be4f1d9"
+KernelAbstractions = "63c18a36-062a-441e-b654-da1e3ab1ce7c"
+Requires = "ae029012-a4dd-5104-9daa-d747884805df"
+
+
+[compat]
+Cassette = "0.3"
+KernelAbstractions = "0.7"
```
Comment (vchuravy) [inline lib/KernelGradients/Project.toml:15]:
```suggestion
KernelAbstractions = "0.7"
Enzyme = "0.6"
```

File: lib/KernelGradients/src/KernelGradients.jl [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+using Requires
+
+@inline function Cassette.overdub(::CPUCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
+    f′ = (args...) -> (Base.@_inline_meta; Cassette.overdub(CPUCTX, f, args...))
+    Enzyme.autodiff(f′, args...)
+end
+
+function Enzyme.autodiff(kernel::Kernel{<:Any, <:Any, <:Any, Fun}) where Fun
+    function df(ctx, args...)
+        Enzyme.autodiff_no_cassette(kernel.f, ctx, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/KernelGradients.jl:15]:
```suggestion
        Enzyme.autodiff(kernel.f, ctx, args...)
```

File: lib/KernelGradients/src/KernelGradients.jl [hunk 8]
```diff
@@ -0,0 +1,25 @@
+module KernelGradients
+
+import KernelAbstractions: Kernel, CPUCTX, CPUCtx
+import Cassette
+import Enzyme
+using Requires
+
+@inline function Cassette.overdub(::CPUCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/KernelGradients.jl:8]:
```suggestion
@inline function Cassette.overdub(::CPUCtx, ::typeof(Enzyme.autodiff), f, args...)
```

File: lib/KernelGradients/src/cuda.jl [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import .CUDAKernels: CUDACtx, CUDACTX
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::CUDACtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/cuda.jl:5]:
```suggestion
@inline function Cassette.overdub(::CUDACtx, ::typeof(Enzyme.autodiff), f, args...)
```

File: lib/KernelGradients/src/roc.jl [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import .ROCKernels: ROCCTX, ROCCtx
+import Cassette
+import Enzyme
+
+@inline function Cassette.overdub(::ROCCtx, ::typeof(Enzyme.autodiff_no_cassette), f, args...)
```
Comment (vchuravy) [inline lib/KernelGradients/src/roc.jl:5]:
```suggestion
@inline function Cassette.overdub(::ROCCtx, ::typeof(Enzyme.autodiff), f, args...)
```

**Comment by vchuravy** [comment]
bors try

**Comment by jpsamaroo** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors try

**Comment by vchuravy** [comment]
bors status

**Comment by vchuravy** [comment]
bors help

**Comment by vchuravy** [comment]
bors try

**Pull request merged**